File: Week 2/control_flow.rs
//! WEEK 2 — Rust: Control Flow & Loops
//! Covers: if/else, match, loop/while/for, break/continue, iterators
//! Equivalent to Java Week 2.

/// Sum even and odd digits — O(d) time, O(1) space
fn sum_even_odd_digits(mut n: i32) -> (i32, i32) {
    let (mut even_sum, mut odd_sum) = (0, 0);
    n = n.abs();
    while n > 0 {
        let digit = n % 10;
        if digit % 2 == 0 { even_sum += digit; }
        else { odd_sum += digit; }
        n /= 10;
    }
    (even_sum, odd_sum)
}

/// All factors — O(sqrt(n)) time
fn factors(n: i32) -> Vec<i32> {
    let mut result = Vec::new();
    let mut i = 1;
    while i * i <= n {
        if n % i == 0 {
            result.push(i);
            if i != n / i { result.push(n / i); }
        }
        i += 1;
    }
    result.sort();
    result
}

/// Power x^n — O(n) time
fn power(x: i64, n: u32) -> i64 {
    let mut result: i64 = 1;
    for _ in 0..n {
        result *= x;
    }
    result
}

fn main() {
    // --- If / Else ---
    // In Rust, if/else is an EXPRESSION (returns a value!)
    let x = 42;
    let category = if x > 100 { "Big" }
                   else if x > 10 { "Medium" }
                   else { "Small" };
    println!("{} is {}", x, category);

    let status = if x % 2 == 0 { "even" } else { "odd" };
    println!("{} is {}", x, status);

    // --- Match (Rust's powerful switch) ---
    println!("\n--- Match ---");
    let day = 3;
    let day_name = match day {
        1 => "Monday",
        2 => "Tuesday",
        3 => "Wednesday",
        4 => "Thursday",
        5 => "Friday",
        6 | 7 => "Weekend",   // or pattern
        _ => "Invalid",        // default (required — must be exhaustive!)
    };
    println!("Day {}: {}", day, day_name);

    // Match with ranges
    let score = 85;
    let grade = match score {
        90..=100 => "A",
        80..=89 => "B",
        70..=79 => "C",
        _ => "F",
    };
    println!("Score {}: Grade {}", score, grade);

    // --- While Loop ---
    println!("\n--- Multiplication Table (while) ---");
    let n = 7;
    let mut i = 1;
    while i <= 10 {
        println!("{} x {} = {}", n, i, n * i);
        i += 1; // Rust has no ++ operator
    }

    // --- Loop (infinite loop with break) ---
    println!("\n--- Loop with break ---");
    let mut count = 0;
    let result = loop {
        count += 1;
        if count == 5 {
            break count * 10; // loop can return a value via break!
        }
    };
    println!("Loop returned: {}", result); // 50

    // --- For Loop ---
    // Rust uses iterators, not C-style for loops
    println!("\n--- For Loop ---");
    print!("1 to 5: ");
    for i in 1..=5 { print!("{} ", i); } // ..= is inclusive range
    println!();

    print!("5 to 1: ");
    for i in (1..=5).rev() { print!("{} ", i); }
    println!();

    print!("Even 0-10: ");
    for i in (0..=10).step_by(2) { print!("{} ", i); }
    println!();

    // --- Even/Odd Digit Sum ---
    println!("\n--- Even/Odd Digit Sum ---");
    for num in [1234, 9876, 555] {
        let (e, o) = sum_even_odd_digits(num);
        println!("{}: even={}, odd={}", num, e, o);
    }

    // --- Factors ---
    println!("\n--- Factors ---");
    for num in [12, 28, 7] {
        println!("Factors of {}: {:?}", num, factors(num));
    }

    // --- Power ---
    println!("\n--- Power ---");
    println!("2^10 = {}", power(2, 10));
    println!("3^5 = {}", power(3, 5));
    // Rust also has: 2_i64.pow(10)

    // --- F to C ---
    println!("\n--- F to C ---");
    println!("{:>5} | {:>8}", "F", "C");
    let mut f = 32;
    while f <= 212 {
        let c = (f as f64 - 32.0) * 5.0 / 9.0;
        println!("{:>5} | {:>8.2}", f, c);
        f += 20;
    }

    // --- Break / Continue ---
    println!("\n--- First prime after 20 ---");
    'outer: for n in 21..100 {
        let mut is_prime = true;
        let mut i = 2;
        while i * i <= n {
            if n % i == 0 { is_prime = false; break; }
            i += 1;
        }
        if is_prime {
            println!("{}", n);
            break 'outer; // labeled break (Rust feature)
        }
    }

    print!("Non-multiples of 3 (1-15): ");
    for i in 1..=15 {
        if i % 3 == 0 { continue; }
        print!("{} ", i);
    }
    println!();

    // --- Iterators (Rust-specific — very powerful) ---
    println!("\n--- Iterators ---");
    let squares: Vec<i32> = (0..10).map(|x| x * x).collect();
    println!("Squares: {:?}", squares);

    let evens: Vec<i32> = (0..20).filter(|x| x % 2 == 0).collect();
    println!("Evens: {:?}", evens);

    let sum: i32 = (1..=100).sum();
    println!("Sum 1..100: {}", sum); // 5050
}

File: Week 2/cpp/control_flow.cpp
/*
 * WEEK 2 — C++ Control Flow & Loops
 * Covers: if/else, while/do-while/for, break/continue, switch
 * Equivalent to Java Week 2.
 */

#include <iostream>
#include <cmath>
#include <vector>
using namespace std;

// Sum even and odd digits — O(d) time, O(1) space
pair<int, int> sumEvenOddDigits(int n) {
    int evenSum = 0, oddSum = 0;
    n = abs(n);
    while (n > 0) {
        int digit = n % 10;
        if (digit % 2 == 0) evenSum += digit;
        else oddSum += digit;
        n /= 10;
    }
    return {evenSum, oddSum};
}

// All factors — O(sqrt(n)) time
vector<int> factors(int n) {
    vector<int> result;
    for (int i = 1; i * i <= n; i++) {
        if (n % i == 0) {
            result.push_back(i);
            if (i != n / i) result.push_back(n / i);
        }
    }
    sort(result.begin(), result.end());
    return result;
}

// Power x^n — O(n) time
long long power(int x, int n) {
    long long result = 1;
    for (int i = 0; i < n; i++) result *= x;
    return result;
}

int main() {
    // --- If / Else ---
    int x = 42;
    if (x > 100) cout << "Big" << endl;
    else if (x > 10) cout << "Medium" << endl;
    else cout << "Small" << endl;

    // Ternary
    string status = (x % 2 == 0) ? "even" : "odd";
    cout << x << " is " << status << endl;

    // --- While Loop: Multiplication Table ---
    cout << "\n--- While Loop ---" << endl;
    int n = 7, i = 1;
    while (i <= 10) {
        cout << n << " x " << i << " = " << n * i << endl;
        i++;
    }

    // --- For Loop ---
    cout << "\n--- For Loop ---" << endl;
    cout << "1 to 5: ";
    for (int i = 1; i <= 5; i++) cout << i << " ";
    cout << endl;

    cout << "5 to 1: ";
    for (int i = 5; i >= 1; i--) cout << i << " ";
    cout << endl;

    cout << "Even 0-10: ";
    for (int i = 0; i <= 10; i += 2) cout << i << " ";
    cout << endl;

    // --- Sum Even/Odd Digits ---
    cout << "\n--- Even/Odd Digit Sum ---" << endl;
    int nums[] = {1234, 9876, 555};
    for (int num : nums) {
        auto [e, o] = sumEvenOddDigits(num); // C++17 structured bindings
        cout << num << ": even=" << e << ", odd=" << o << endl;
    }

    // --- Factors ---
    cout << "\n--- Factors ---" << endl;
    for (int num : {12, 28, 7}) {
        cout << "Factors of " << num << ": ";
        for (int f : factors(num)) cout << f << " ";
        cout << endl;
    }

    // --- Power ---
    cout << "\n--- Power ---" << endl;
    cout << "2^10 = " << power(2, 10) << endl;
    cout << "3^5 = " << power(3, 5) << endl;

    // --- F to C table ---
    cout << "\n--- F to C ---" << endl;
    printf("%5s | %8s\n", "F", "C");
    for (int f = 32; f <= 212; f += 20) {
        double c = (f - 32) * 5.0 / 9.0;
        printf("%5d | %8.2f\n", f, c);
    }

    // --- Break / Continue ---
    cout << "\n--- Break: first prime after 20 ---" << endl;
    for (int n = 21; n < 100; n++) {
        bool isPrime = true;
        for (int i = 2; i * i <= n; i++) {
            if (n % i == 0) { isPrime = false; break; }
        }
        if (isPrime) { cout << n << endl; break; }
    }

    cout << "Non-multiples of 3 (1-15): ";
    for (int i = 1; i <= 15; i++) {
        if (i % 3 == 0) continue;
        cout << i << " ";
    }
    cout << endl;

    // --- Switch ---
    cout << "\n--- Switch ---" << endl;
    int day = 3;
    switch (day) {
        case 1: cout << "Monday" << endl; break;
        case 2: cout << "Tuesday" << endl; break;
        case 3: cout << "Wednesday" << endl; break;
        default: cout << "Other" << endl;
    }

    return 0;
}
